File: clients/python/amenolink/_connection.py
from typing import Callable
from ._connection_manager import connection_manager, ConnectionStatus

type ConnectionStatusListener = Callable[[ConnectionStatus], None]

MAX_RECONNECT_ATTEMPTS = 5
CONNECTION_TIMEOUT_SECONDS = 5.0


class Connection:
    def __init__(self) -> None:
        self._listeners: set[ConnectionStatusListener] = set()

    def subscribe(self, listener: ConnectionStatusListener) -> None:
        self._listeners.add(listener)

    def unsubscribe(self, listener: ConnectionStatusListener) -> None:
        self._listeners.discard(listener)

    def unsubscribe_all(self) -> None:
        self._listeners.clear()

    def _on_connection_changed(self, status: ConnectionStatus) -> None:
        for listener in list(self._listeners):
            listener(status)

    def connect(self) -> None:
        connection_manager.connect(
            on_status_change=self._on_connection_changed,
            max_attempts=MAX_RECONNECT_ATTEMPTS,
            timeout_seconds=CONNECTION_TIMEOUT_SECONDS,
        )

    def disconnect(self) -> None:
        connection_manager.disconnect()


connection = Connection()

File: clients/typescript/src/connection.ts
import { ConnectionManager, connectionManager, ConnectionStatus } from './connection_manager';
import { Connection as IConnection } from './interfaces';

export class Connection implements IConnection {
    private static readonly MAX_RECONNECT_ATTEMPTS = 5;
    private static readonly CONNECTION_TIMEOUT_SECONDS = 5.0;

    private readonly listeners: Set<(status: ConnectionStatus) => void> = new Set();
    private readonly manager: ConnectionManager;

    constructor(manager: ConnectionManager = connectionManager) {
        this.manager = manager;
    }

    public subscribe(listener: (status: ConnectionStatus) => void): void {
        this.listeners.add(listener);
    }

    public unsubscribe(listener: (status: ConnectionStatus) => void): void {
        this.listeners.delete(listener);
    }

    public unsubscribeAll(): void {
        this.listeners.clear();
    }

    private onConnectionChanged = (status: ConnectionStatus): void => {
        for (const listener of Array.from(this.listeners))
            listener(status);
    };

    public connect(): Promise<void> {
        return this.manager.connect({
            onStatusChange: this.onConnectionChanged,
            maxAttempts: Connection.MAX_RECONNECT_ATTEMPTS,
            timeoutSeconds: Connection.CONNECTION_TIMEOUT_SECONDS
        });
    }

    public disconnect(): Promise<void> {
        return this.manager.disconnect();
    }
}

export const connection = new Connection();
